config/kernel-get-user-pages.m4

```diff
@@ -0,0 +1,171 @@
+dnl #       int write, int force, struct page **pages)
+dnl # 4.8 API Change
+dnl # long get_user_pages_unlocked(unsigned long start,
+dnl #       unsigned long nr_pages, int write, int force, struct page **page)
+dnl # 4.9 API Change
+dnl # long get_user_pages_unlocked(usigned long start, int nr_pages,
+dnl #       struct page **pages, unsigned int gup_flags)
+dnl #
+dnl #
+dnl # In earlier kernels (< 4.0) get_user_pages() is available
+dnl # 
+dnl # int get_user_pages(struct task_struct *tsk, struct mm_struct *mm,
+dnl #       unsigned long start, int nr_pages, int write, int force,
+dnl #       struct_page **pages, struct vm_area_struct **vmas)
+dnl #
+dnl # 4.6 API Change
+dnl # long get_user_pages(unsigned long start, unsigned long nr_pages,
+dnl #       unsigned int gup_flags, struct page **pages,
+dnl #       struct vm_area_struct **vmas)
+dnl #
+
+dnl#
+dnl# Check available get_user_pages/_unlocked interfaces.
+dnl#
+AC_DEFUN([ZFS_AC_KERNEL_SRC_GET_USER_PAGES], [
+	ZFS_LINUX_TEST_SRC([get_user_pages_unlocked_gup_flags], [
+		#include <linux/mm.h>
+	], [
+		unsigned long start = 0;
+		unsigned long nr_pages = 1;
+		unsigned int gup_flags = 0;
+		struct page **pages = NULL;
+		long ret __attribute__ ((unused));
+		ret = get_user_pages_unlocked(start, nr_pages, pages, gup_flags);
+	])
+
+	ZFS_LINUX_TEST_SRC([get_user_pages_unlocked_write_flag], [
+		#include <linux/mm.h>
+	], [
+		unsigned long start = 0;
+		unsigned long nr_pages = 1;
+		int write = 0;
+		int force = 0;
+		long ret __attribute__ ((unused));
+		struct page **pages = NULL;
+		ret = get_user_pages_unlocked(start, nr_pages, write, force, pages);
+	])
+			
+
+	ZFS_LINUX_TEST_SRC([get_user_pages_unlocked_task_struct], [
+		#include <linux/mm.h>
+	], [
+		struct task_struct *tsk = NULL;
+		struct mm_struct *mm = NULL;
+		unsigned long start = 0;
+		unsigned long nr_pages = 1;
+		int write = 0;
+		int force = 0;
+		struct page **pages = NULL;
+		long ret __attribute__ ((unused));
+		ret = get_user_pages_unlocked(tsk, mm, start, nr_pages, write,
+		    force, pages);
+	])
+	
+	ZFS_LINUX_TEST_SRC([get_user_pages_gup_flags], [
+		#include <linux/mm.h>
+	], [
+		struct vm_area_struct **vmas = NULL;
+		unsigned long start = 0;
+		unsigned long nr_pages = 1;
+		unsigned int gup_flags = 0;
+		struct page **pages = NULL;
+		long ret __attribute__ ((unused));
+		ret = get_user_pagees(start, nr_pages, gup_flags, pages, vmas);
+	])
+
+	ZFS_LINUX_TEST_SRC([get_user_pages_task_struct], [
+		#include <linux/mm.h>
+	], [
+		struct task_struct *tsk = NULL;
+		struct mm_struct *mm = NULL;
+		struct vm_area_struct **vmas = NULL;
+		unsigned long start = 0;
+		unsigned long nr_pages = 1;
+		int write = 0;
+		int force = 0;
+		struct page **pages = NULL;
+		int ret __attribute__ ((unused));
+		ret = get_user_pages(tsk, mm, start, nr_pages, write,
+		    force, pages, vmas);
+	])
+])
+
+dnl #
+dnl # Supported get_user_pages/_unlocked interfaces checked newest to oldest.
```

Comment:
So, looking at this again, the 4.6 kernel check is kinda pointless on my part. We should never hit it. The idea here was to use get_user_pages() in case an OS distro didn't provide the get_user_pages_unlocked() API call. The get_user_pages() check with the task_struct can be used all the way back to kernel 2.6, so we can just have the single get_user_pages() check and if the kernel is older than 4.0. At what kernel version do we cut off checks?

Reply:
The oldest kernel we need to support is the 3.10 kernel.

Reply:
Cool, I will go ahead and remove the extra get_user_pages() check here as it is unnecessary.